### ISSUE TITLE
Ctrl-Backspace crashes editor with multiple cursors on same line

### ISSUE THREAD

**Issue by mackenziedg** [godotengine/godot#73476]
### Godot version

v4.0.rc2.official

### System information

Window 10

### Issue description

If multiple cursors are selecting two words on the same line in the code editor, pressing Control + Backspace crashes the program. This only seems to happen if they are on the same line (different lines are fine as far as I've seen). The steps below have crashed 100% of the time for me with error

```
ERROR: FATAL: Index p_index = 10 is out of bounds (size() = 9).
   at: get (./core/templates/cowdata.h:155)
```

### Steps to reproduce

Paste the following anywhere into the code editor

```
fun crash
```

Place a cursor at the end of each word

```
fun| crash|
```

Press control + backspace.

The following will not crash

```
var| fun| = crash
```

### Minimal reproduction project

N/A

**Comment by MewPurPur** [comment]
I've managed to get text editor to crash a few times in the last few days - not only like this - and it always pointed to the cow data.

**Comment by akien-mga** [comment]
> and it always pointed to the cow data.

There's nothing to fix in cowdata, aside from porting Godot from C++ to Rust - this is a simple out of bounds indexing in a vector. It can indeed happen in tons of different situations but the fix won't be in cowdata but in the calling code that wrongly indexes a vector.

**Comment by MewPurPur** [comment]
Cows are never wrong! I'm just saying it says if it points at the same thing, then it must be a different flavor of the same bug :)

Looking into it.

**Comment by akien-mga** [comment]
What I meant is that if you see this error in different circumstances:
```
ERROR: FATAL: Index p_index = <something> is out of bounds (size() = <something>).
   at: get (./core/templates/cowdata.h:155)
```
There's no reason to think it's actually the same bug. Any Vector out of bounds read anywhere in the codebase would end up on this exact line. The only differences would be the values for `p_index` and `size()`.

**Comment by AndyBarcia** [comment]
I think that the bug is somewhere in this code

https://github.com/godotengine/godot/blob/f84479fe5a0a3c34c1d0217527ae08dd24543243/scene/gui/text_edit.cpp#L2627-L2693

**Comment by MewPurPur** [comment]
I swore I told Paulb about this bug in the original PR that introduced multi-carets. But apparently it's been here since time immemorial.

**Comment by MewPurPur** [comment]
Oh man, there's more to this than meets the eye. Unless someone can fix the crash, the 4.1 (potential 4.0.x) milestone seems fair. Meanwhile I'm trying to rewrite `_do_backspace()` completely
